lib/Conversion/TorchToTMTensor/TorchToTMTensor.cpp

```diff
@@ -0,0 +1,208 @@
+
+    // Finding the maximum value in the input tensor.
+    SmallVector<int64_t> maxTensorSizes;
+    ValueTensorType maxTensorType = ValueTensorType::get(
+        context, llvm::makeArrayRef(maxTensorSizes),
+        op.self().getType().cast<ValueTensorType>().getDtype());
+    Value maxTensor = rewriter.create<AtenMaxOp>(loc, maxTensorType, op.self());
+    TypeConverter *typeConverter = getTypeConverter();
+    maxTensor = typeConverter->materializeTargetConversion(
+        rewriter, loc, typeConverter->convertType(maxTensor.getType()),
+        maxTensor);
+
+    // `maxTensor` is a 0-d tensor, extracting its only element and
+    // storing it in `maxInput`.
+    Value maxInput = rewriter.create<tensor::ExtractOp>(loc, maxTensor);
+
+    // Creating a tm_tensor.scatter op with the following mapping:
+    // 1.) `input` tensor maps to the indices in scatter op. `input` is
+    // expanded from 1-d to 2-d, and its element type is set to i32 as required
+    // for the scatter op.
+    // 2.) `updates` is a 1-d dummy tensor with the size equivalent to the
+    // `input`.
+    // 3.) `bincount` a 1-d tensor maps to the original in scatter op
+    // with size equal to the max(max(input) + 1, minlength).
+    SmallVector<int64_t> expandedInputSizes{inputType.getShape()[0], 1};
+    ValueTensorType expandInputType = ValueTensorType::get(
+        context, llvm::makeArrayRef(expandedInputSizes),
+        op.self().getType().cast<ValueTensorType>().getDtype());
+    Value torchCstOne = rewriter.create<Torch::ConstantIntOp>(
+        loc, rewriter.getI64IntegerAttr(1));
+    Value expandedInputTensor = rewriter.create<AtenUnsqueezeOp>(
+        loc, expandInputType, op.self(), torchCstOne);
+
+    // Converting the input element type from i64 to i32.
+    Value indices = convertTensorToDtype(
+        rewriter, loc, expandedInputTensor,
+        mlir::IntegerType::get(context, 32, mlir::IntegerType::Signed));
+    indices = typeConverter->materializeTargetConversion(
+        rewriter, loc, typeConverter->convertType(indices.getType()), indices);
+
+    Type resultElemType = typeConverter->convertType(op->getResult(0).getType())
+                              .cast<RankedTensorType>()
+                              .getElementType();
+
+    SmallVector<Value, 1> inputSize;
+    if (!inputType.hasStaticShape())
+      inputSize = getTensorSizesUntilDim(rewriter, loc, input, 0);
+    Value updatesTensor = rewriter.create<linalg::InitTensorOp>(
+        loc, inputSize, inputType.getShape(), resultElemType);
+
+    Value constantZero = rewriter.create<arith::ConstantOp>(
+        loc, rewriter.getZeroAttr(inputType.getElementType()));
+    Value constantOne = rewriter.create<arith::ConstantIntOp>(
+        loc, 1, resultElemType.getIntOrFloatBitWidth());
+
+    // Bincount size = max(max(input) + 1, minlength)
+    Value maxInputPlusOne =
+        rewriter.create<arith::AddIOp>(loc, maxInput, constantOne);
```

Comment:
Can you verify if this would work if the input is `torch.int32`? The `maxInput` and `cst1` would have different bit width in such case. It's probably easier to just convert the input to `int64` upfront.